lib/codecs/src/encoding/format/syslog.rs

```diff
@@ -0,0 +1,577 @@
+    pub fn new(conf: &SyslogSerializerConfig) -> Self {
+        Self { config: conf.clone() }
+    }
+}
+
+impl Encoder<Event> for SyslogSerializer {
+    type Error = vector_common::Error;
+
+    fn encode(&mut self, event: Event, buffer: &mut BytesMut) -> Result<(), Self::Error> {
+        if let Event::Log(log_event) = event {
+            let syslog_message = ConfigDecanter::new(log_event).decant_config(&self.config.syslog);
+
+            let vec = syslog_message
+                .encode(&self.config.syslog.rfc)
+                .as_bytes()
+                .to_vec();
+            buffer.put_slice(&vec);
+        }
+
+        Ok(())
+    }
+}
+
+// Adapts a `LogEvent` into a `SyslogMessage` based on config from `SyslogSerializerOptions`:
+// - Splits off the responsibility of encoding logic to `SyslogMessage` (which is not dependent upon Vector types).
+// - Majority of methods are only needed to support the `decant_config()` operation.
+struct ConfigDecanter {
+    log: LogEvent,
+}
+
+impl ConfigDecanter {
+    fn new(log: LogEvent) -> Self {
+        Self {
+            log,
+        }
+    }
+
+    fn decant_config(&self, config: &SyslogSerializerOptions) -> SyslogMessage {
+        let y = |v| self.replace_if_proxied_opt(v);
+        let app_name = y(&config.app_name).unwrap_or("vector".to_owned());
+        let proc_id = y(&config.proc_id);
+        let msg_id = y(&config.msg_id);
+
+        SyslogMessage {
+            pri: Pri {
+                facility: config.facility,
+                severity: config.severity,
+            },
+            timestamp: self.get_timestamp(),
+            hostname: self.value_by_key("hostname"),
+            tag: Tag {
+                app_name,
+                proc_id,
+                msg_id,
+            },
+            structured_data: self.get_structured_data(),
+            message: self.get_message(&config),
+        }
+    }
+
+    fn replace_if_proxied_opt(&self, value: &Option<String>) -> Option<String> {
+        value.as_ref().and_then(|v| self.replace_if_proxied(v))
+    }
+
+    // When the value has the expected prefix, perform a lookup for a field key without that prefix part.
+    // A failed lookup returns `None`, while a value without the prefix uses the config value as-is.
+    //
+    // Q: Why `$.message.` as the prefix? (Appears to be JSONPath syntax?)
+    // NOTE: Originally named in PR as: `get_field_or_config()`
+    fn replace_if_proxied(&self, value: &str) -> Option<String> {
+        value
+            .strip_prefix("$.message.")
+            .map_or(
+                Some(value.to_owned()),
+                |field_key| self.value_by_key(field_key),
+            )
+    }
+
+    fn value_by_key(&self, field_key: &str) -> Option<String> {
+        self.log.get(field_key).map(|field_value| {
+            field_value.to_string_lossy().to_string()
+        })
+    }
+
+    fn get_structured_data(&self) -> Option<StructuredData> {
+        self.log.get("structured_data")
```

Comment:
For the Vector namespace, (when `log_namespace: true`) we want to get this by meaning instead of name:

```suggestion
  match self.log.namespace() {
       LogNamespace::Vector => self.log.get_by_meaning("structured_data")
       LogNamespace::Legacy => self.log.get("structured_data")
  }
```